numpy/core/src/multiarray/datetime.c

```diff
@@ -1761,6 +1762,9 @@ parse_datetime_unit_from_string(char *str, Py_ssize_t len, char *metastr)
                 return NPY_FR_as;
         }
     }
+    else if (allow_generic && len == 7 && !strncmp(str, "generic", 7)) {
```

Comment:
If it is zero, the fix probably belongs elsewhere. I think as is too many things may be interpreted as generic. Could maybe be tested by trying to generate datetimes with invalid types with length > 2 and != 7.